### PR TITLE
New vmctl module to interface with the OpenBSD VMM hypervisor

### PR DESCRIPTION
**Pull request by jasperla**
### What does this PR do?

Provide a new module that allows Salt to interact with the OpenBSD VMM hypervisor.

### Previous Behavior

NA

### New Behavior

This new module implements the following functionality:

- create a new disk image
- load a new configuration file
- reload the default configuration
- stop VMs
- reset VMM or one of the subsystems
- define new VMs and start existing VMs
- request status from all or one VM(s)

### Tests written?

No

### Commits signed with GPG?

Yes

### CODE REVIEW

File: salt/modules/vmctl.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
+# -*- coding: utf-8 -*-
+# $OpenBSD$
+'''
+Manage vms running on the OpenBSD VMM hypervisor using vmctl(8).
+
```
Comment (rallytime) [inline salt/modules/vmctl.py]: A versionadded for `Fluorine` will be needed for this new file.

File: salt/modules/vmctl.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+'''
+
+from __future__ import absolute_import
+
+# Import python libs
+import logging
+import re
+
+# Imoprt salt libs:
+import salt.utils
```
Comment (rallytime) [inline salt/modules/vmctl.py]: We're moving away from bare `salt.utils` imports. All of the functions there have been moved over into their own files. Can you update this import to import the individual files needed `import salt.utils.x`? (And of course update the references in the function calls below.)

The `salt.utils.__init__.py` file should have the up-to-date paths in the deprecation warnings for each function for reference.

**Review by rallytime — CHANGES_REQUESTED** [review]
@jasperla Thanks again for these additions! Keep 'em coming!

I have a couple of comments here. This will also need some unit tests added as well before we can get it in.

File: salt/modules/vmctl.py (hunk 1 above)
Comment (jasperla) [reply salt/modules/vmctl.py]: Check, I'll it.

File: salt/modules/vmctl.py (hunk 2 above)
Comment (jasperla) [reply salt/modules/vmctl.py]: There was actually only one (`salt.utils.path`). I'll adjust it.

**Comment by jasperla** [comment]
I've now added tests too.

File: salt/modules/vmctl.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+        salt '*' vmctl.status           # to list all VMs
+        salt '*' vmctl.status name=web1 # to get a single VM
+    '''
+    ret = {}
+    cmd = ['vmctl', 'status']
+
+    if name:
+        cmd.append(name)
+    else:
+        cmd.append(id)
```
Comment (eradman) [inline salt/modules/vmctl.py]: Should be `if id:`

**Review by eradman — COMMENTED** [review]
If no argument is passed to `vmctl.status` then this function mistakenly filters VMs by the name `None`

    [INFO    ] Executing command ['vmctl', 'status', None] in directory '/home/eradman'
    [TRACE   ] stdout:    ID   PID VCPUS  MAXMEM  CURMEM     TTY        OWNER NAME
    [DEBUG   ] LazyLoaded nested.output
    [TRACE   ] data = {'local': {}}

**Review by eradman — CHANGES_REQUESTED** [review]
There is an edge case not handled here when `vmctl` reports that the VM is in the state "stopping"

File: salt/modules/vmctl.py (hunk 3 above)
Comment (jasperla) [reply salt/modules/vmctl.py]: Good point, I had accidentally changed that. It's fixed now.

**Comment by jasperla** [comment]
@eradman I've fixed the edge case too now. Turns out it's a bug in vmctl where it doesn't output the status line for a VM that is stopping when the status is requested for the VM by name (works by id). i.e. `vmctl stop foo && vmctl status foo` emits the header but nothing else.

File: salt/modules/vmctl.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    switch:
+        Add a network interface that is attached to the specified
+        virtual switch on the host.
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt '*' vmctl.start 2   # start VM with id 2
+        salt '*' vmctl.start name=web1 boothpath='/bsd.rd' nics=2 memory=512M disk='/disk.img'
```
Comment (eradman) [inline salt/modules/vmctl.py]: s/boothpath/bootpath/

**Review by eradman — CHANGES_REQUESTED** [review]
@jasperla: previous fixes work. There is another flaw in the way `vmctl` reports status that causes some confusion. If I start a VM with a name that is already used the output is unclear:

    $ doas salt-call --local vmctl.start name=web1 bootpath='/bsd.rd' nics=2 memory=512M disk='/tmp/disk.img'
    local:
        ----------
        changes:
            True
        console:
            None

The output from `vmctl` is not much better

    $ doas vmctl start web1 -i 2 -b /bsd.rd -m 512M -d /tmp/disk.img
    vmctl: start vm command failed: Operation not permitted
    $ echo $?
    0

Not sure what this Salt module should do--perhaps we need to parse the status line for this VM?

    (saltenv) $ doas vmctl status web1
       ID   PID VCPUS  MAXMEM  CURMEM     TTY        OWNER NAME
       12 54520     1    512M   93.3M   ttypj         root web1
     VCPU:  0 STATE: STOPPED

**Comment by jasperla** [comment]
@eradman thanks for pointing out that bug in `vmctl`, I'll look into that.

I've now changed `start()` to return:

```
local:
    ----------
    changes:
        False
    comment:
        VM already exists and cannot be redefined
    console:
        None
```

So `changes` is now `False` and a `comment` is populated in case the VM that one tries to define already exists.

I do seem to have broken the `test_start_new_vm` as `start()` now calls `status()` whose return value I cannot seem to mock properly to return `{}`.

**Comment by eradman** [comment]
Functionally this change seems to be good!

**Review by rallytime — APPROVED** [review]

**Comment by rallytime** [comment]
re-run py

**Comment by rallytime** [comment]
re-run py

**Comment by jasperla** [comment]
Looking better now :)

**Pull request merged**